pkg/processor/schemaregistry/avro/schema.go

```diff
@@ -0,0 +1,202 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package avro
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/hamba/avro/v2"
+)
+
+// ExtractSchema uses reflection to extract an Avro schema from v.
+func ExtractSchema(v any) (avro.Schema, error) {
+	return reflectInternal([]string{"record"}, reflect.ValueOf(v), reflect.TypeOf(v))
+}
+
+// SortFields is a utility for tests to ensure the schemas can be compared.
+func SortFields(s avro.Schema) {
+	rs, ok := s.(*avro.RecordSchema)
+	if !ok {
+		return
+	}
+	fields := rs.Fields()
+	sort.Slice(fields, func(i, j int) bool {
+		return fields[i].Name() < fields[j].Name()
+	})
+	for i := range fields {
+		SortFields(fields[i].Type())
+	}
+}
+
+//nolint:gocyclo // reflection requires a huge switch, it's fine
+func reflectInternal(path []string, v reflect.Value, t reflect.Type) (avro.Schema, error) {
+	if t == nil {
+		return nil, nil // untyped nil
+	}
+	switch t.Kind() {
+	case reflect.Bool:
+		return avro.NewPrimitiveSchema(avro.Boolean, nil), nil
+	case reflect.Int, reflect.Int64, reflect.Uint32:
+		return avro.NewPrimitiveSchema(avro.Long, nil), nil
+	case reflect.Int32, reflect.Int16, reflect.Uint16, reflect.Int8, reflect.Uint8:
+		return avro.NewPrimitiveSchema(avro.Int, nil), nil
+	case reflect.Float32:
+		return avro.NewPrimitiveSchema(avro.Float, nil), nil
+	case reflect.Float64:
+		return avro.NewPrimitiveSchema(avro.Double, nil), nil
+	case reflect.String:
+		return avro.NewPrimitiveSchema(avro.String, nil), nil
+	case reflect.Pointer:
+		s, err := reflectInternal(path, v.Elem(), t.Elem())
+		if err != nil {
+			return nil, err
+		}
+
+		var schemas avro.Schemas
+		if us, ok := s.(*avro.UnionSchema); ok && us.Nullable() {
+			// it's already a nullable schema
+			return s, nil
+		} else if ok {
+			// take types from union schema
+			schemas = us.Types()
+		} else if s.Type() != avro.Null {
+			// non-nil type
+			schemas = avro.Schemas{s}
+		}
+
+		s, err = avro.NewUnionSchema(append(schemas, &avro.NullSchema{}))
+		if err != nil {
+			return nil, err
+		}
+
+		return s, nil
+	case reflect.Interface:
+		if !v.IsValid() || v.IsNil() {
+			return nil, nil // can't get a schema for this
+		}
+		return reflectInternal(path, v.Elem(), v.Elem().Type())
+	case reflect.Array, reflect.Slice:
+		if t.Elem().Kind() == reflect.Uint8 {
+			return avro.NewPrimitiveSchema(avro.Bytes, nil), nil
+		}
+
+		var elemValue reflect.Value
+		if v.Len() > 0 {
+			elemValue = v.Index(0)
```

Comment:
Good point, currently we make an assumption they are all the same type. I can try to take the type info from the slice type, although the issue is that I'm not sure if the type info of slices is preserved when the connector sends us the data. Also, if we are receiving raw data (e.g. JSON) and are unmarshaling it into structured data, we will almost certainly lack the type info for slices.

To sum up - the question is if we see `[]any{1,2,3}`, should we assume that the type is always `int`, even though theoretically the slice could contain any type? 🤔

Reply:
Can you remind me, how would schema updates work? I'm wondering if we can use a union type here and update it if needed. E.g. if we have `[]any{"hi", 123}` we have a union type with the string and int types. A new type might be added in the future, in which case we update the schema (if possible).

Reply:
We don't start with an existing schema, we simply extract it from the data that we have in the current record. There is no update in that sense. We try to register the extracted schema and either get back an ID of an existing schema (ideally it was already cached, so we didn't even send a request), or the new updated schema gets stored in the registry and we get back a new ID.

Reply:
@lovromazgon so for the new, updated, schema -- how is it connected to the previous one, for the records from the same source? Thinking aloud, a source _may_ return homogenous data, so we can't assume that all records belong to the same schema, so it has to be a new schema always?

Reply:
The processor will have an option to configure the subject for the extracted schema, the schema registry figures out the version (if it already contains the exact schema under that subject it uses that, otherwise creates a new version). The subject can be static or taken from the record (e.g. from metadata), so we should be able to handle different records.